src/internal/storage/kv/lru_cache.go

```diff
@@ -0,0 +1,107 @@
+package kv
+
+import (
+	"context"
+	"sync"
+
+	"github.com/hashicorp/golang-lru/v2/simplelru"
+	"github.com/pachyderm/pachyderm/v2/src/internal/log"
+	"github.com/pachyderm/pachyderm/v2/src/internal/stream"
+	"go.uber.org/zap"
+)
+
+var _ Store = &LRUCache{}
+
+type LRUCache struct {
+	slow, fast Store
+	mu         sync.RWMutex
+	cache      simplelru.LRU[string, struct{}]
+
+	ctx context.Context
+}
+
+func NewLRUCache(slow, fast Store, size int) *LRUCache {
+	c := &LRUCache{
+		slow: slow,
+		fast: fast,
+	}
+	cache, err := simplelru.NewLRU[string, struct{}](size, c.onEvict)
+	if err != nil {
+		panic(err)
+	}
+	c.cache = *cache
+	return c
+}
+
+func (c *LRUCache) Get(ctx context.Context, key []byte, buf []byte) (int, error) {
+	// note that we don't need a lock here because stores are thread-safe and Put/Deletes are atomic.
+	n, err := c.fast.Get(ctx, key, buf)
+	if err == nil {
+		// cache hit
+		return n, nil
+	}
+	// cache miss
+	n, err = c.slow.Get(ctx, key, buf)
+	if err != nil {
+		return 0, err
+	}
+	if err := c.putFast(ctx, key, buf[:n]); err != nil {
+		log.Error(ctx, "kv.LRUCache could not put in fast layer", zap.Error(err))
+	}
+	return n, nil
+}
+
+func (c *LRUCache) Put(ctx context.Context, key, value []byte) error {
+	return c.slow.Put(ctx, key, value)
+}
+
+func (c *LRUCache) Exists(ctx context.Context, key []byte) (bool, error) {
```

Comment:
I think this is the correct behavior for an Exists method, but we need to double check that we use postgres as the source of truth for whether we can skip an upload.

Reply:
We do check with postgres in `chunk/client.go`